runtime/ops/web_worker.rs

```diff
@@ -2,40 +2,113 @@
 
 use crate::web_worker::WebWorkerHandle;
 use crate::web_worker::WorkerEvent;
+use deno_core::error::custom_error;
+use deno_core::error::generic_error;
+use deno_core::error::AnyError;
 use deno_core::futures::channel::mpsc;
+use deno_core::futures::stream::StreamExt;
+use deno_core::serde::Deserialize;
+use deno_core::serde_json;
 use deno_core::serde_json::{json, Value};
+use deno_core::ZeroCopyBuf;
+use std::sync::Arc;
+use tokio::sync::Mutex as AsyncMutex;
+
+#[derive(Deserialize)]
+struct HostUnhandledErrorArgs {
+  message: String,
+}
 
 pub fn init(
   rt: &mut deno_core::JsRuntime,
   sender: mpsc::Sender<WorkerEvent>,
-  handle: WebWorkerHandle,
+  receiver: Arc<AsyncMutex<mpsc::Receiver<Box<[u8]>>>>,
+  webworker_handle: WebWorkerHandle,
 ) {
-  // Post message to host as guest worker.
-  let sender_ = sender.clone();
-  super::reg_json_sync(
+  super::reg_bin_async(
+    rt,
+    "op_worker_get_message",
+    move |_state, _args, _bufs| {
+      let receiver_clone = receiver.clone();
+      op_worker_get_message(receiver_clone)
+    },
+  );
+
+  let sender_clone1 = sender.clone();
+  super::reg_bin_sync(
     rt,
     "op_worker_post_message",
-    move |_state, _args: Value, bufs| {
-      assert_eq!(bufs.len(), 1, "Invalid number of arguments");
-      let msg_buf: Box<[u8]> = (*bufs[0]).into();
-      sender_
-        .clone()
-        .try_send(WorkerEvent::Message(msg_buf))
-        .expect("Failed to post message to host");
-      Ok(json!({}))
+    move |_state, _args, bufs| {
+      let sender_clone = sender_clone1.clone();
+      op_worker_post_message(bufs, sender_clone)
     },
   );
 
-  // Notify host that guest worker closes.
+  let sender_clone1 = sender.clone();
+  super::reg_bin_sync(rt, "op_worker_close", move |_state, _args, _bufs| {
+    let sender_clone = sender_clone1.clone();
+    op_worker_close(&webworker_handle, sender_clone)
+  });
+
   super::reg_json_sync(
     rt,
-    "op_worker_close",
-    move |_state, _args: Value, _bufs| {
-      // Notify parent that we're finished
-      sender.clone().close_channel();
-      // Terminate execution of current worker
-      handle.terminate();
-      Ok(json!({}))
+    "op_worker_unhandled_error",
+    move |_state, args, _zero_copy| {
+      let sender_clone = sender.clone();
+      op_worker_unhandled_error(args, sender_clone)
     },
   );
 }
+
+fn op_worker_unhandled_error(
+  args: Value,
+  mut sender: mpsc::Sender<WorkerEvent>,
+) -> Result<Value, AnyError> {
+  let args: HostUnhandledErrorArgs = serde_json::from_value(args)?;
+  sender
+    .try_send(WorkerEvent::Error(generic_error(args.message)))
+    .expect("Failed to propagate error event to parent worker");
+  Ok(json!(true))
+}
+
+/// Get message from host as worker
+async fn op_worker_get_message(
+  receiver_ref: Arc<AsyncMutex<mpsc::Receiver<Box<[u8]>>>>,
+) -> Result<Vec<u8>, AnyError> {
+  let mut receiver = receiver_ref.try_lock()?;
```

Comment:
It should be just `lock`; if there are multiple `op_worker_get_message` ops dispatched it should be a panic indicating problem with internal code.

Reply:
Here this is used too, is this different or should it change too?
https://github.com/denoland/deno/blob/26f7a3f185881c12081eb6d3cedbf9e637376d86/runtime/web_worker.rs#L76-L77